src/framework/linalg/eigen_psd.hpp

```diff
@@ -22,17 +22,168 @@
  * Returns the eigenvalues and eigenvectors of a Hermitian
  * positive semi-definite (PSD) matrix.
  * @param psd_matrix: The Hermitian PSD matrix.
- * @param eigenvalues: The eignevalues of the matrix.
- * @param eigenvectors: The eigenvectors of the matrix.
+ * @param eigenvalues: On output: the eignevalues of the matrix (input is overwritten)
+ * @param eigenvectors: On input: a copy of psd_matrix, on output: the eigenvectors of the matrix.
  *
- * @returns: true.
+ * @returns: void
  */
-template <class float_t>
-void eigensystem_psd(const matrix<std::complex<float_t>>& psd_matrix,
-               /* out */ std::vector<std::complex<float_t>>& eigenvalues,
-               /* out */ std::vector<std::vector<std::complex<float_t>>>& eigenvectors){
+template <class T>
+void eigensystem_psd(const matrix<std::complex<T>>& psd_matrix,
+               /* out */ std::vector<T>& eigenvalues,
+               /* out */ matrix<std::complex<T>>& eigenvectors);
 
+//template <>
+//void eigensystem_psd(const matrix<std::complex<float>>& psd_matrix,
+//               /* out */ std::vector<float>& eigenvalues,
+//               /* out */ matrix<std::complex<float>>& eigenvectors); 
+//template <>
+//void eigensystem_psd(const matrix<std::complex<double>>& psd_matrix,
+//               /* out */ std::vector<double>& eigenvalues,
+//               /* out */ matrix<std::complex<double>>& eigenvectors); 
 
+template <class T>
+void eigensystem_psd(const matrix<std::complex<T>>& psd_matrix,
+               /* out */ std::vector<T>& eigenvalues,
+               /* out */ matrix<std::complex<T>>& eigenvectors) {}
+
+template <>
+void eigensystem_psd(const matrix<std::complex<double>>& psd_matrix,
+               /* out */ std::vector<double>& eigenvalues,
+               /* out */ matrix<std::complex<double>>& eigenvectors) {
+#ifdef DEBUG
+  if ( psd_matrix.getRows() != psd_matrix.getCols() ) {
```

Comment:
this should probably raise an exception if the checks aren't met - rather than what would happen right now (likely a seg fault in the lapack functions - an actual qiskit job would probably never make it down here with incorrect dimensions but still an exception is probably the best way to go

Reply:
just so I don't forget it should go if ( getRows == getCols ) {do work and return} else { log and throw }

Reply:
Actually, it's considered a good practice to "return early" (search for early return pattern).
We can save an else block and it's easier to see the preconditions.
So:
```c++
if(getRows != getCols){
  return / throw !
}
```